dialer.go

```diff
@@ -50,97 +62,124 @@ func NewDialer(opts *DialerOpts) Dialer {
+	pool                  BufferPool
+	cipherCode            Cipher
+	serverPublicKey       *rsa.PublicKey
+	muNumLivePending      sync.Mutex
+	numLivePending        int
+	liveSessions          chan sessionInf
 }
 
 func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialStream(dial)
+	return d.DialContext(context.Background(), dial)
 }
 
-func (d *dialer) DialStream(dial DialFN) (Stream, error) {
-	d.mx.Lock()
-	current := d.current
-	idsExhausted := false
-	if d.id > d.maxStreamPerConn {
-		log.Debug("Exhausted maximum allowed IDs on one physical connection, will open new connection")
-		idsExhausted = true
-		d.id = 0
-	}
-	idled := false
-	if d.idleInterval > 0 {
-		now := time.Now()
-		idled = now.Sub(d.lastDialed) > d.idleInterval
-		if idled {
-			log.Debugf("No new connections in %v, will start new session", d.idleInterval)
-		}
-		d.lastDialed = now
+func (d *dialer) DialContext(ctx context.Context, dial DialFN) (net.Conn, error) {
+	s, err := d.getOrCreateSession(ctx, dial)
+	if err != nil {
+		return nil, err
 	}
+	c := s.CreateStream()
+	d.liveSessions <- s
+	return c, nil
+}
 
-	// TODO: support pooling of connections (i.e. keep multiple physical connections in flight)
-	if current == nil || idsExhausted || idled {
-		if current != nil {
-			// Need to run this on a goroutine because we call back to sessionClosed()
-			// via the session's beforeClose callback, which needs the mutex that's
-			// already locked here.
-			go current.markDefunct()
-		}
+func (d *dialer) getNumLivePending() int {
+	d.muNumLivePending.Lock()
+	numLivePending := d.numLivePending
+	d.muNumLivePending.Unlock()
+	return numLivePending
+}
 
-		var err error
-		current, err = d.startSession(dial)
-		if err != nil {
-			d.mx.Unlock()
-			return nil, err
+func (d *dialer) getOrCreateSession(ctx context.Context, dial DialFN) (sessionInf, error) {
+	newSession := func(cap int) {
+		d.muNumLivePending.Lock()
+		if d.numLivePending >= cap {
+			d.muNumLivePending.Unlock()
+			return
+		}
+		d.numLivePending++
+		d.muNumLivePending.Unlock()
+		go func() {
+			s, err := d.startSession(dial)
+			if err != nil {
+				d.muNumLivePending.Lock()
+				d.numLivePending--
+				d.muNumLivePending.Unlock()
+				return
+			}
+			d.liveSessions <- s
+		}()
+	}
+	for {
+		select {
+		case s := <-d.liveSessions:
+			if s.AllowNewStream(d.maxStreamsPerConn, d.idleInterval) {
+				return s, nil
+			}
+			d.muNumLivePending.Lock()
+			d.numLivePending--
+			d.muNumLivePending.Unlock()
+			s.MarkDefunct()
+			newSession(1)
+		case <-time.After(d.redialSessionInterval):
+			newSession(d.maxLiveConns)
+		case <-ctx.Done():
+			return nil, errors.New("No session available")
 		}
 	}
-	id := d.id
-	d.id++
-	d.mx.Unlock()
-
-	c, _ := current.getOrCreateStream(id)
-	return c, nil
 }
 
 func (d *dialer) EMARTT() time.Duration {
 	var rtt time.Duration
-	d.mx.Lock()
-	current := d.current
-	d.mx.Unlock()
+	current := <-d.liveSessions
 	if current != nil {
 		rtt = current.EMARTT()
```

Comment:
If necessary, we can change the code to calculate EMARTT in the dialer, instead of the session. I omitted it as most of time there's only one session.